src/ModelContextProtocol/McpEndpointExtensions.cs

```diff
@@ -1,13 +1,145 @@
+    /// <param name="parameters">Object representing the request parameters.</param>
+    /// <param name="parametersTypeInfo">The type information for request parameter serialization.</param>
+    /// <param name="resultTypeInfo">The type information for request parameter deserialization.</param>
+    /// <param name="requestId">The request id for the request.</param>
+    /// <param name="cancellationToken">A token to cancel the operation.</param>
+    /// <returns>A task that represents the asynchronous operation. The task result contains the deserialized result.</returns>
+    public static async Task<TResult> SendRequestAsync<TParameters, TResult>(
+        this IMcpEndpoint endpoint,
+        string method,
+        TParameters parameters,
+        JsonTypeInfo<TParameters> parametersTypeInfo,
+        JsonTypeInfo<TResult> resultTypeInfo,
+        RequestId? requestId = null,
+        CancellationToken cancellationToken = default)
+        where TResult : notnull
+    {
+        Throw.IfNull(endpoint);
+        Throw.IfNullOrWhiteSpace(method);
+        Throw.IfNull(parametersTypeInfo);
+        Throw.IfNull(resultTypeInfo);
+
+        JsonRpcRequest jsonRpcRequest = new()
+        {
+            Method = method,
+            Params = JsonSerializer.SerializeToNode(parameters, parametersTypeInfo),
+        };
+
+        if (requestId is { } id)
+        {
+            jsonRpcRequest.Id = id;
+        }
+
+        JsonRpcResponse response = await endpoint.SendRequestAsync(jsonRpcRequest, cancellationToken).ConfigureAwait(false);
+        return JsonSerializer.Deserialize(response.Result, resultTypeInfo) ?? throw new JsonException("Unexpected JSON result in response.");
+    }
+
+    /// <summary>
+    /// Sends a JSON-RPC request and attempts to deserialize the result to <typeparamref name="TResult"/>.
+    /// </summary>
+    /// <typeparam name="TParameters">The type of the request parameters to serialize from.</typeparam>
+    /// <typeparam name="TResult">The type of the result to deserialize to.</typeparam>
+    /// <param name="endpoint">The MCP client or server instance.</param>
+    /// <param name="method">The JSON-RPC method name to invoke.</param>
+    /// <param name="parameters">Object representing the request parameters.</param>
+    /// <param name="serializerOptions">The options governing request serialization.</param>
+    /// <param name="requestId">The request id for the request.</param>
+    /// <param name="cancellationToken">A token to cancel the operation.</param>
+    /// <returns>A task that represents the asynchronous operation. The task result contains the deserialized result.</returns>
+    public static Task<TResult> SendRequestAsync<TParameters, TResult>(
+        this IMcpEndpoint endpoint,
+        string method,
+        TParameters parameters,
+        JsonSerializerOptions? serializerOptions = null,
+        RequestId? requestId = null,
```

Comment:
Nit: I'd consider reordering these, so that requestId comes before serializerOptions. Serializer options is more infrastructure and less likely to be supplied, whereas request ID is more closely associated with the method/parameters.

Reply:
It follows the ordering of the overloads accepting JsonTypeInfo where it must follow these values because they're non-optional.

Reply:
But this overload doesn't have JsonTypeInfos. I'm not understanding.

Reply:
The order of parameters for both overloads is roughly `(method, parameters, Json(TypeInfo|SerializerOptions), RequestId? = null)`. I'm pointing out that we can only change the order for the JSO overload making the order inconsistent.

Reply:
Following feedback from @halter73 I'm inclined to just mark the JTI overloads as internal for now.

Reply:
Yep yep